### ISSUE TITLE
ValueError: `Field` default cannot be set in `Annotated` for discriminated unions

### ISSUE THREAD

**Issue by cfxegbert** [samuelcolvin/pydantic#3991]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.9.0
            pydantic compiled: True
                 install path: /Users/*****/Library/Caches/pypoetry/virtualenvs/mei-imf-z9_q_NMv-py3.10/lib/python3.10/site-packages/pydantic
               python version: 3.10.2 (main, Feb  2 2022, 08:42:42) [Clang 13.0.0 (clang-1300.0.29.3)]
                     platform: macOS-11.6.5-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

```py
"""Bug: ValueError: `Field` default cannot be set in `Annotated` for similar classes"""

from typing import Literal, Annotated, Union, List, Optional
from uuid import UUID

from pydantic import BaseModel, Extra, Field


class BaseResource(BaseModel, extra=Extra.forbid):
    """BaseResourceType as defined by SMPTE ST 2067-3:2016"""
    id: UUID


class TrackFileResource(BaseResource):
    """TrackFileResourceType as defined by SMPTE ST 2067-3:2016"""
    type: Literal["TrackFileResourceType"]
    sourceEncoding: UUID


class MarkerResource(BaseResource):
    """MarkerResourceType as defined by SMPTE ST 2067-3:2016"""
    type: Literal["MarkerResourceType"]
    label: str


Resource = Annotated[Union[TrackFileResource, MarkerResource], Field(discriminator='type')]


class SequenceType(BaseModel, extra=Extra.forbid):
    """Sequence as defined in SMPTE ST 2067-3:2016"""
    id: UUID
    trackId: UUID
    resourceList: List[Resource]


class CDPSequenceType(BaseModel, extra=Extra.forbid):
    """CDPSequence as defined in SMPTE ST 2067-2:2020"""
    id: UUID
    trackId: UUID
    parentTrackID: UUID
    resourceList: List[Resource]


class Sequence(BaseModel, extra=Extra.forbid):
    """Sequence including SMPTE ST 2067-3:2016 and SMPTE ST 2067-2:2020"""
    mainImageSequence: Optional[SequenceType]
    cdpSequence: Optional[CDPSequenceType]
```
With the above code when I try to import I get
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/*****/amazon/git/mei/mei_imf/mei_imf/model/test_bug.py", line 35, in <module>
    class CDPSequenceType(BaseModel, extra=Extra.forbid):
  File "pydantic/main.py", line 204, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 488, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 419, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 534, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 728, in pydantic.fields.ModelField._type_analysis
  File "pydantic/fields.py", line 776, in pydantic.fields.ModelField._create_sub_type
  File "pydantic/fields.py", line 451, in pydantic.fields.ModelField._get_field_info
ValueError: `Field` default cannot be set in `Annotated` for '_resourceList'
```
If I change `CDPSequenceType` to use `SequenceType` as a base it works
```py
class CDPSequenceType(SequenceType):
    """CDPSequence as defined in SMPTE ST 2067-2:2020"""
    parentTrackID: UUID
```

**Comment by accelleon** [comment]
Hmm this might have something to do with how the types are handled internally. Curiously this works:
```python
Resource = Annotated[Union[TrackFileResource, MarkerResource], Field(discriminator='type')]
Resource2 = Annotated[Union[TrackFileResource, MarkerResource], Field(discriminator='type')]


class SequenceType(BaseModel, extra=Extra.forbid):
    """Sequence as defined in SMPTE ST 2067-3:2016"""
    id: UUID
    trackId: UUID
    resourceList: List[Resource]


class CDPSequenceType(BaseModel, extra=Extra.forbid):
    """CDPSequence as defined in SMPTE ST 2067-2:2020"""
    id: UUID
    trackId: UUID
    parentTrackID: UUID
    resourceList: List[Resource2]
```

**Comment by cfxegbert** [comment]
This was going to be in a library.  I'll just have to document the problem out of existence.  It would be nice to somehow have a discriminated union with having to use a Field.  Field seems to be a catchall for everything right now.  Actually have some sort of `DiscrimatedUnion` type that you could use to annotate the union instead of a Field.  The Field(discriminator='foo') could be kept for backwards compatibility.

**Comment by accelleon** [comment]
If your intention is to retain a public "type hint" for the `Union` I reckon this serves your purpose:
```python
Resource = Union[TrackFileResource, MarkerResource]

class SequenceType(BaseModel, extra=Extra.forbid):
    """Sequence as defined in SMPTE ST 2067-3:2016"""
    id: UUID
    trackId: UUID
    resourceList: List[Annotated[Resource, Field(discriminator='type')]]

class CDPSequenceType(BaseModel, extra=Extra.forbid):
    """CDPSequence as defined in SMPTE ST 2067-2:2020"""
    id: UUID
    trackId: UUID
    parentTrackID: UUID
    resourceList: List[Annotated[Resource, Field(discriminator='type')]]
```

Realistically the `Annotated[..., Field(discriminator=...)]` is only for telling pydantic how to differentiate the two unions to avoid type coercion from UUID to str in your specific example.

**Comment by peterbplanetly** [comment]
I've also been struggling with this problem. In particular, the problem is even worse when working with nested discriminated unions. Toy example:

```py
import typing as t
import pydantic


class SpringerSpaniel(pydantic.BaseModel):
    animal_type: t.Literal["dog"] = "dog"
    dog_type: t.Literal["springer_spaniel"] = "springer_spaniel"


class GoldenRetriever(pydantic.BaseModel):
    animal_type: t.Literal["dog"] = "dog"
    dog_type: t.Literal["golden_retriever"] = "golden_retriever"


Dog = t.Annotated[t.Union[SpringerSpaniel, GoldenRetriever], pydantic.Field(discriminator="dog_type")]


class Tabby(pydantic.BaseModel):
    animal_type: t.Literal["cat"] = "cat"
    cat_type: t.Literal["tabby"] = "tabby"


class Bengal(pydantic.BaseModel):
    animal_type: t.Literal["cat"] = "cat"
    cat_type: t.Literal["bengal"] = "bengal"


Cat = t.Annotated[t.Union[Tabby, Bengal], pydantic.Field(discriminator="cat_type")]

Animal = t.Annotated[t.Union[Dog, Cat], pydantic.Field(discriminator="animal_type")]

#

class User(pydantic.BaseModel):
    pet: Animal


class Mating(pydantic.BaseModel):
    a1: Animal
    a2: Animal
```

In this case it is not possible to just use `Annotated` inline as suggested here https://github.com/samuelcolvin/pydantic/issues/3991#issuecomment-1099852863, since we want to use the final union `Animal` _and_ the intermediate unions `Dog`, `Cat` in other parts of the code.

---

> I don't think this is really a bug per-say. 

I agree that is expected, given how `Field` works. But, I would probably say it is still a bug. Unless I misunderstand it, this makes it pretty impossible to reuse intermediate unions in nested discriminated unions.

 >  It would be nice to somehow have a discriminated union with[out] having to use a Field. Field seems to be a catchall for everything right now. Actually have some sort of DiscrimatedUnion type that you could use to annotate the union instead of a Field. The Field(discriminator='foo') could be kept for backwards compatibility.

nice idea i think, maybe an API change like that can help.

**Comment by peterbplanetly** [comment]
pretty sure this is the same issue as https://github.com/samuelcolvin/pydantic/issues/3714?

**Comment by berzi** [comment]
I'd like to point out that this limitation prevents useful annotate types from being reused.

Say I want to have a type like:
```py
NonNegativeFloat = Annotated[float, Field(ge=0)]
```

I can't actually use it unless I stick just the part after the equals on the type of a field, *and copy-paste it on any other field that needs it* instead of reusing just the `NonNegativeFloat` reference, as is otherwise usual to do with any other complex type hint.

You might argue that I could just use `confloat(ge=0)` but:

1. Say my annotation doesn't just need `ge` but a custom metadata, I can only apply that with an Annotated Field.
2. Mypy can't deal with functions as type hints, so I have to use Annotated. 

More easily (re-)usable type hints means more type-hinted code, which means better code everywhere. this issue makes the difference between having just `int` and having more accurate information, especially on large and complex codebases.

I realise the fix to this might not be easy but I think the matter needs to at the very least be considered something to be changed eventually and not a closed issue.

**Comment by jaraco** [comment]
In https://github.com/jaraco/inflect/issues/169, a user reports having encountered a similar error message. I'm unsure if it's the same issue. I can't replicate the issue locally using the same code. Can someone take a peek and help me understand if I should file a separate issue and if there is anything the inflect project can do to address the issue for users that encounter it? I'd also appreciate any advice on how I can replicate the user's experience.

**Comment by samuelcolvin** [comment]
This will be fixed in v2, very much doubt we are brave enough to try and fix this in v1.10.
